doc/main/scala/lms/core/backend.md

```diff
@@ -1,45 +1,35 @@
-# LMS IR
+The `lms/core/backend.scala` file contains several important aspects of the LMS IR. In `object Backend`, we define the basic building blocks of the IR and effect-related data structures. In `class GraphBuilder`, `class GraphBuilderOpt` and `case class Graph`, we introduce how new nodes are added into a graph IR and how effects are handled and used to calculate dependencies. We also define `Phase` as a framework for graph-to-graph transformation and introduce a dead code elimination (DCE) phase `DeadCodeElimCG`.
 
-In the file lms/core/backend.scala, the core LMS IR is defined in `object Backend`.
+# Backend
+The core LMS IR is defined in `object Backend`. We first introduce the basic building blocks of the IR:
 
 ``` scala
 abstract class Def // Definition: used in right-hand-side of all nodes
 abstract class Exp extends Def
 case class Sym(n: Int) extends Exp   // Symbol
 case class Const(x: Any) extends Exp // Constant
 case class Block(in: List[Sym], res: Exp, ein: Sym, eff: EffectSummary) extends Def
+
 case class Node(n: Sym, op: String, rhs: List[Def], eff: EffectSummary)
 ```
 
-From this class hierarchy, we can see that the basic unit of the LMS IR is `Exp`,
-which can be `Const(0)`, `Const("hello")`, or `Sym(2)`, which really means `x2` in the
-generated code. We can combine basic units into `Node`, such as
+`Def` represents a definition, which is used in the RHS of all nodes. A definition can be an `Exp` or a `Block`. `Exp` can be a `Sym` or a `Const`. `Sym` represents symbols in the IR identified by a unique integer. For example, `Sym(2)` represents `x2` in the generated code. `Const` represents immediate values, such as integers (e.g. `Const(0)`), strings (e.g. `Const(hello)`), etc.
 
+`Node` represents statements in the IR. The first field `n` of a node is a `Sym` corresponding to the LHS of a statement. The second field `op` is the operator of the statement. We use strings as operators since they permit easy extension of various kinds of nodes such as `-`, `print`, etc. The third field is the list of operands (RHS) of a node. Let's ignore the last field `EffectSummary` for now, which will be covered later. For example, the statement `val x3 = x2 + 1` would become
```

Comment:
```suggestion
`Node` represents statements in the IR. The first field `n` of a node is a `Sym` corresponding to the identity of a statement. The second field `op` is the operator of the statement. We use strings as operators since they permit easy extension of various kinds of nodes such as `-`, `print`, etc. The third field is the list of operands (RHS) of a node. Let's ignore the last field `EffectSummary` for now, which will be covered later. For example, the statement `val x3 = x2 + 1` would become
```